third/cns/src/man/kinit.1
.\" Copyright 1989 by the Massachusetts Institute of Technology.
.\"
.\" For copying and distribution information,
.\" please see the file <mit-copyright.h>.
.\"
.TH KINIT 1 "Kerberos Version 4.0" "MIT Project Athena"
.SH NAME
kinit \- Kerberos login utility
.SH SYNOPSIS
.B kinit
[
.B \-irvls
]
[
.I user[.instance][@realm]
]
.SH DESCRIPTION
The
.I kinit
command is used to login to the
Kerberos
authentication and authorization system.
Note that only registered
Kerberos
users (called
.I principals
) can use the
Kerberos
system.
For information about registering as a
Kerberos
user,
see the
.I kerberos(1)
manual page.
.PP
The optional argument is a Kerberos principal name, which may
simply be a user name.  It may also specify an instance and/or
a realm name.  If no argument is supplied, 
.I kinit
prompts for a user name.
.PP
.I kinit
then prompts for your Kerberos password,
and tries to authenticate your login with the Kerberos
server for the local realm (or for the realm you specified).
.PP
If
Kerberos
authenticates the login attempt,
.I kinit
retrieves your initial ticket and puts it in the ticket file specified by
your KRBTKFILE environment variable.
If this variable is undefined,
your ticket will be stored in the
.IR /tmp
directory,
in the file
.I tktuid ,
where
.I uid
specifies your user identification number.
.PP
Make sure you use the
.I kdestroy
command to destroy any active tickets before you end your login session.
You may want to put the
.I kdestroy
command in your
.I \.logout
file so that your tickets will be destroyed automatically when you logout.
.PP
The options to
.I kinit
are as follows:
.TP 7
.B \-i
.I kinit
prompts you for a
Kerberos
instance.  It's usually easier to specify the entire
principal name on the command line.
.TP
.B \-r
.I kinit
prompts you for a
Kerberos
realm.
This option lets you authenticate yourself with a remote
Kerberos
server.  It's usually easier to specify the entire
principal name on the command line.
.TP
.B \-v
Verbose mode.
.I kinit
prints the name of the ticket file used, and
a status message indicating the success or failure of
your login attempt.
.TP
.B \-l
.I kinit
prompts you for a ticket lifetime in minutes.  Due to protocol
restrictions in Kerberos Version 4, this value must be between 5 and
1275 minutes.
.TP
.B \-s
Instead of using a password to prove who you are,
.I kinit
prompts you with a challenge to be entered into a
Digital Pathways SNK-4 authentication token.  (This requires
that your secret SNK-4 key has been added to the Kerberos database
using the 
.I "kadmin snk"
command.)  Example:
.nofill

 % kinit -s username
 Challenge: 123456
 Response:12345678
 % klist
 Ticket file:	/tmp/tkt86
 Principal:	username.+SNK4@REALM.ORG

 Issued           Expires          Principal
 Dec 31 18:35:30  Dec 32 04:35:30
                              krbtgt.REALM.ORG@REALM.ORG

 or

 % kinit -s username
 Challenge #1: 123456
 Response #1:12345678
 Challenge #2: 789012
 Response #2:78901234
 %
.fill
.SH SEE ALSO
.PP
kerberos(1), kdestroy(1), klist(1), kadmin(8)
.SH AUTHORS
Steve Miller, MIT Project Athena/Digital Equipment Corp.
.br
Clifford Neuman, MIT Project Athena


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/cns/src/kuser/kinit.c
/*
 * kinit.c
 *
 * Copyright 1987, 1988 by the Massachusetts Institute of Technology. 
 *
 * For copying and distribution information, please see the file
 * <mit-copyright.h>. 
 *
 * Routine to initialize user to Kerberos.  Prompts optionally for
 * user, instance and realm.  Authenticates user and gets a ticket
 * for the Kerberos ticket-granting service for future use. 
 *
 * Options are: 
 *   -i[instance]
 *   -r[realm]
 *   -v[erbose]
 *   -l[ifetime]
 * and you can supply a full blown principal.instance@realm arg also.
 */

#include "mit-copyright.h"
#include <stdio.h>
#define	DEFINE_SOCKADDR
#include "krb.h"
#include <string.h>

#include <krb5.h>

#define	LIFE	DEFAULT_TKT_LIFE   /* lifetime of ticket in 5-minute units */

char   *progname;

void
get_input(s, size, stream)
char *s;
int size;
FILE *stream;
{
	char *p;

	if (fgets(s, size, stream) == NULL)
	  exit(1);
	if ( (p = strchr(s, '\n')) != NULL)
		*p = '\0';
}


static char
hex_scan_nybble(c)
     char c;
{
  if (c >= '0' && c <= '9')
    return c - '0';
  if (c >= 'A' && c <= 'F')
    return c - 'A' + 10;
  if (c >= 'a' && c <= 'f')
    return c - 'a' + 10;
  return -1;
}

/* returns: NULL for ok, pointer to error string for bad input */
static char*
hex_scan_four_bytes(out, in)
     char *out;
     char *in;
{
  int i;
  char c, c1;
  for (i=0; i<8; i++) {
    if(!in[i]) return "not enough input";
    c = hex_scan_nybble(in[i]);
    if(c<0) return "invalid digit";
    c1 = c; i++;
    if(!in[i]) return "not enough input";
    c = hex_scan_nybble(in[i]);
    if(c<0) return "invalid digit";
    *out++ = (c1 << 4) + c;
  }
  switch(in[i]) {
  case 0:
  case '\r':
  case '\n':
    return 0;
  default:
    return "extra characters at end of input";
  }
}

/*
 * This routine takes v4 kinit parameters and performs a V5 kinit.
 * 
 * name, instance, realm is the v4 principal information
 *
 * lifetime is the v4 lifetime (i.e., in units of 5 minutes)
 * 
 * password is the password
 *
 * ret_cache_name is an optional output argument in case the caller
 * wants to know the name of the actual V5 credentials cache (to put
 * into the KRB5_ENV_CCNAME environment variable)
 *
 * etext is a mandatory output variable which is filled in with
 * additional explanatory text in case of an error.
 * 
 */
krb5_error_code do_v5_kinit(name, instance, realm, lifetime, password,
			    ret_cache_name, etext)
	char	*name;
	char	*instance;
	char	*realm;
	int	lifetime;
	char	*password;
	char	**ret_cache_name;
	char	**etext;
{
	krb5_context context;
	krb5_error_code retval;
	krb5_principal me = 0, server = 0;
	krb5_ccache ccache = NULL;
	krb5_creds my_creds;
	krb5_timestamp now;
	krb5_flags options = KDC_OPT_FORWARDABLE | KDC_OPT_PROXIABLE;

	char *cache_name;

	*etext = 0;
	if (ret_cache_name)
		*ret_cache_name = 0;
	memset(&my_creds, 0, sizeof(my_creds));

	retval = krb5_init_context(&context);
	if (retval)
		return retval;

	cache_name = krb5_cc_default_name(context);
	krb5_init_ets(context);

	retval = krb5_425_conv_principal(context, name, instance, realm, &me);
	if (retval) {
		*etext = "while converting V4 principal";
		goto cleanup;
	}

	retval = krb5_cc_resolve(context, cache_name, &ccache);
	if (retval) {
		*etext = "while resolving ccache";
		goto cleanup;
	}

	retval = krb5_cc_initialize(context, ccache, me);
	if (retval) {
		*etext = "while initializing cache";
		goto cleanup;
	}

	retval = krb5_build_principal_ext(context, &server,
					  krb5_princ_realm(context,
							   me)->length,
					  krb5_princ_realm(context, me)->data,
					  KRB5_TGS_NAME_SIZE, KRB5_TGS_NAME,
					  krb5_princ_realm(context,
							   me)->length,
					  krb5_princ_realm(context, me)->data,
					  0);
	if (retval)  {
		*etext = "while building server name";
		goto cleanup;
	}

	retval = krb5_timeofday(context, &now);
	if (retval) {
		*etext = "while getting time of day";
		goto cleanup;
	}

	my_creds.client = me;
	my_creds.server = server;
	my_creds.times.starttime = 0;
	my_creds.times.endtime = now + lifetime*5*60;
	my_creds.times.renew_till = 0;

	retval = krb5_get_in_tkt_with_password(context, options, NULL, NULL,
					       NULL, password, ccache,
					       &my_creds, NULL);
	if (retval) {
		*etext = "while calling krb5_get_in_tkt_with_password";
		goto cleanup;
	}

	if (ret_cache_name) {
		*ret_cache_name = (char *) malloc(strlen(cache_name)+1);
		if (!*ret_cache_name) {
			retval = ENOMEM;
			goto cleanup;
		}
		strcpy(*ret_cache_name, cache_name);
	}

cleanup:
	if (me)
		krb5_free_principal(context, me);
	if (server)
		krb5_free_principal(context, server);
	if (ccache)
		krb5_cc_close(context, ccache);
	my_creds.client = 0;
	my_creds.server = 0;
	krb5_free_cred_contents(context, &my_creds);
	krb5_free_context(context);
	return retval;
}

int 
main(argc, argv)
    int     argc;
    char   *argv[];
{
    char    aname[ANAME_SZ];
    char    inst[INST_SZ];
    char    realm[REALM_SZ];
    char    password[BUFSIZ];
    char    buf[MAXHOSTNAMELEN];
    char   *username = NULL;
    char   *usernameptr;
    int     iflag, rflag, vflag, lflag, pflag, sflag, lifetime, k_errno;
    register char *cp;
    register i;
 
#if 0
/* This code is for testing time conversion in micro Kerberos ports */
    {
	extern int krb_debug;
	static long time_0 = 0L;
	static long time_1 = 1L;
	static long time_epoch;
	static long time_epoch2;

	krb_debug = 0xFFF;
	time_epoch = win_time_get_epoch();
	time_epoch2 = -time_epoch;
	printf ("STime at zero: %s\n", krb_stime(&time_0));
	printf ("STime at one: %s\n", krb_stime(&time_1));
	printf ("STime at Epoch: %s\n", krb_stime(&time_epoch));
	printf ("STime at -Epoch: %s\n", krb_stime(&time_epoch2));
	printf ("CTime at zero: %s\n", ctime(&time_0));
	printf ("CTime at one: %s\n", ctime(&time_1));
	printf ("CTime at Epoch: %s\n", ctime(&time_epoch));
	printf ("CTime at -Epoch: %s\n", ctime(&time_epoch2));
    }
#endif

    *inst = *realm = '\0';
    iflag = rflag = vflag = lflag = pflag = sflag = 0;
    lifetime = LIFE;

    /* Take last component with either / or \ as path separator.  */
    cp = strrchr(*argv, '/');
    progname = cp ? cp + 1 : *argv;
    cp = strrchr(progname, '\\');
    progname = cp ? cp + 1 : progname;

    while (--argc) {
	if ((*++argv)[0] != '-') {
	    if (username)
		usage();
	    username = *argv;
	    continue;
	}
	for (i = 1; (*argv)[i] != '\0'; i++)
	    switch ((*argv)[i]) {
	    case 'i':		/* Instance */
		++iflag;
		continue;
	    case 'r':		/* Realm */
		++rflag;
		continue;
	    case 'v':		/* Verbose */
		++vflag;
		continue;
	    case 'l':
		++lflag;
		continue;
            case 'p':
                ++pflag;
                continue;
            case 's':
                ++sflag;
                continue;
	    default:
		usage();
		exit(1);
	    }
    }

    krb_start_session ((char *)NULL);

    if (username &&
	(k_errno = kname_parse(aname, inst, realm, username))
	!= KSUCCESS) {
	fprintf(stderr, "%s: %s\n", progname, krb_get_err_text(k_errno));
	iflag = rflag = 1;
	username = NULL;
    }

    buf[0] = '\0';
    if (GETHOSTNAME (buf, MAXHOSTNAMELEN))
      buf[0] = 0;

    if (buf[0])
      printf ("Kerberos initialization on %s\n", buf);
    else
      printf("Kerberos initialization\n");
    if (!username) {
	printf("Kerberos name: ");
	get_input(aname, sizeof(aname), stdin);
	if (!*aname)
	    exit(0);
    }
    /* optional instance */
    if (iflag) {
	printf("Kerberos instance: ");
	get_input(inst, sizeof(inst), stdin);
    }
    if (rflag) {
	printf("Kerberos realm: ");
	get_input(realm, sizeof(realm), stdin);
    }
    if (lflag) {
	 printf("Kerberos ticket lifetime (minutes): ");
	 get_input(buf, sizeof(buf), stdin);
	 lifetime = atoi(buf);
	 if (lifetime < 5)
	      lifetime = 1;
	 else
	      lifetime /= 5;
	 /* This should be changed if the maximum ticket lifetime */
	 /* changes */
	 if (lifetime > 255)
	      lifetime = 255;
    }
    if (!*realm && krb_get_lrealm(realm, 1)) {
	fprintf(stderr, "%s: Error attempting to determine local Kerberos realm name\n", progname);
	exit(1);
    }

 
    if (!sflag) {
#if defined(_WINDOWS) || defined(macintosh)
	printf("Password: ");
	get_input(password, sizeof(password), stdin);
#else
	des_read_pw_string(password, BUFSIZ, "Password: ", 0);
#endif

	if (pflag) {
	    k_errno = krb_get_pw_in_tkt_preauth(aname, inst, realm, "krbtgt",
						realm, lifetime, password);
	} else {
	    k_errno = krb_get_pw_in_tkt(aname, inst, realm, "krbtgt", realm,
					lifetime, password);
	}
	if (k_errno == 0) {
	    krb5_error_code retval;
	    char	*etext;
	    
	    retval =  do_v5_kinit(aname, inst, realm, lifetime, password,
				  0, &etext);
	    memset(password, 0, BUFSIZ);
	    if (retval && retval != KRB5_CONFIG_CANTOPEN &&
		retval != KRB5KRB_AP_ERR_BAD_INTEGRITY) {
		    com_err(argv[0], retval, etext);
	    }
	}
    } else { /* sflag is set, so we do snk4 support instead */
      /* for the SNK4 we have to get the ticket explicitly. */

      KTEXT_ST cip_st;
      KTEXT cip = &cip_st;	/* Returned Ciphertext */
      KTEXT_ST cip2_st;
      KTEXT cip2 = &cip2_st;	/* Returned Ciphertext */
      /* preauth is impossible in this context */
      if (pflag)
	{
	  fprintf(stderr, "%s: preauth not supported with SNK device\n");
	  exit(1);
	}

      /* flag the instance to get the right tickets */
      strcat(inst, "+SNK4");

      if (k_errno = krb_mk_in_tkt_preauth(aname, inst, realm, 
					  "krbtgt", realm,
					  lifetime, (char*)0, 0, cip))
	{
	    fprintf(stderr, "%s: %s for principal %s%s%s@%s\n", progname,
		    krb_get_err_text(k_errno), aname, inst[0]?".":"", inst,
		    realm);
	    exit(1);
	}

      /* here we've got the response. Pull off the challenge. */
      /* decrypt_tkt (user, instance, realm, arg, key_proc, &cip); */
      {
	static char version[] = "cnssnk01";
	des_cblock key1, key2;
	char challenge1[8], challenge2[8];
	char resp1[128], resp2[128];
	des_key_schedule ks;
	char *complain;

	if (memcmp (cip->dat, version, 8)) {
	  /* perhaps try the normal decrypt in this case? */
	  fprintf(stderr, "%s: not an SNK response packet\n", progname);
	  exit(1);
	}
	memcpy(challenge1, cip->dat+8, 8);
	memcpy(challenge2, cip->dat+16, 8);
	
	/* add input checks */
	if (memcmp (challenge1, challenge2, 8)) {
	  complain = NULL;
	  do {
	    if(complain) printf("Bad input: %s\n", complain);
	    printf("Challenge #1: %s\nResponse #1:", challenge1);
	    get_input(resp1, sizeof(resp1), stdin);
	  } while(complain = hex_scan_four_bytes((char*)&key1, resp1));
	  complain = NULL;
	  do {
	    if(complain) printf("Bad input: %s\n", complain);
	    printf("Challenge #2: %s\nResponse #2:", challenge2);
	    get_input(resp2, sizeof(resp2), stdin);
	  } while(complain = hex_scan_four_bytes(4+(char*)&key1, resp2));
	} else {
	  complain = NULL;
	  do {
	    if(complain) printf("Bad input: %s\n", complain);
	    printf("Challenge: %s\nResponse:", challenge1);
	    get_input(resp1, sizeof(resp1), stdin);
	  } while(complain = hex_scan_four_bytes((char*)&key1, resp1));
	  memcpy(4+(char*)&key1, (char*)&key1, 4);
	  strcpy(resp2, resp1);
	}
	des_fixup_key_parity(key1);
	des_key_sched(key1, ks);
	des_ecb_encrypt(key1, key2, ks, 1);
	des_fixup_key_parity(key2);

	des_key_sched(key2,ks);
	pcbc_encrypt(cip->dat+24,cip2->dat,
		     (long) (cip->length-24),ks, key2,0);
	cip2->length = cip->length-24;

      }
    
      /* if this fails, perhaps the user should get another chance
	 at the input? */
      if (k_errno = krb_parse_in_tkt(aname, inst, realm, 
				     "krbtgt", realm, lifetime, cip2))
	{
	  fprintf(stderr, "%s: %s\n", progname, krb_get_err_text(k_errno));
	  exit(1);
	}
    }

    if (vflag) {
	printf("Result from realm %s: ", realm);
	printf("%s\n", krb_get_err_text(k_errno));
    } else if (k_errno) {
        fprintf(stderr, "%s: %s for principal %s%s%s@%s\n", progname,
		krb_get_err_text(k_errno), aname, inst[0]?".":"", inst, realm);
	exit(1);
    }

#if 0
    /* More testing code for micro ports */
 
    usernameptr = krb_get_default_user();
    printf ("Default user is `%s'.\n", usernameptr);

    k_errno = krb_set_default_user ("Bogon");
    printf ("Set default user result is %d\n", k_errno);

    printf ("Now default user is `%s'.\n", krb_get_default_user());
    
    k_errno = krb_set_default_user (usernameptr);

    /* Test credentials caching code 
     * Tests are order dependent and depend on the results of previous tests.
     */
    {
    int numcreds;
    CREDENTIALS cr;
    char    aname[ANAME_SZ];
    char    inst[INST_SZ];
    char    realm[REALM_SZ];
    
    /* krb_get_num_cred 
     * returns the number of credentials in the default cache.
     */
    numcreds = krb_get_num_cred();
    printf ("After getting tgt...\n");
    printf ("The number of credentials in the default cache:%d.\n", numcreds);
    
    /* krb_get_tf_fullname
     * For the ticket file name input, return the principal's name, 
     * instance, and realm.	 Currently the ticket file name is 
     * unused because there is just one default ticket file/cache.
     */
     
     printf ("krb_get_tf_fullname returns %d.\n", 
     	krb_get_tf_fullname ((char*)0, aname, inst, realm));
	 printf ("principal's name: %s, instance: %s, realm: %s. \n", 
	 	aname, inst, realm);		
    
	/* krb_get_nth_cred 
	 * returns service name, service instance and realm of the nth credential. 
	 */
	 
	printf ("Get the service name, instance, and realm of the last credential.\n");
	printf ("krb_get_nth_cred returns %d.\n", 
			krb_get_nth_cred(aname, inst, realm, numcreds));
	printf ("service name: %s, instance: %s, realm: %s.\n",
    		aname, inst, realm);	
    		
    /* krb_get_cred 
     * fills in a cred with info from the first cred in the default cache
     * with matching sname, inst, and realm 
     */
    printf ("Make a new cred and fill it in with info from the cred\n");
    printf ("in the default cache with matching service name, inst and realm.\n");
    
    printf ("krb_get_cred returns: %d.\n", krb_get_cred(aname, inst, realm, &cr));

    printf ("service: %s, instance: %s, realm: %s\n", cr.service, cr.instance, cr.realm);
  	printf ("pname: %s, pinst: %s.\n", cr.pname, cr.pinst);
    
    /* krb_save_credentials 
     * adds a new credential to the default cache, using information input
     */ 
    printf ("Add a new cred to the cache, using mostly values from the cred made above\n");
    printf ("krb_save_credentials returns: %d.\n",
    		 krb_save_credentials ("bogus", cr.instance, cr.realm, 
    							   cr.session, cr.lifetime, cr.kvno, 
    							   &cr.ticket_st, cr.issue_date));
    							   
    printf ("number of credentials in the cache: %d.\n", krb_get_num_cred());			  
    							      
    /* krb_delete_cred 	
     * deletes the first credential in the default cache 
     * with matching service name, instance and realm. 
     */		
     printf ("Delete the first cred in the cache with matching service name, inst, realm\n");
     printf ("krb_delete_cred returns: %d.\n",
     		  krb_delete_cred ("bogus", inst, realm));
     
     printf ("number of credentials in the cache: %d.\n", krb_get_num_cred());			  
    } /* end credentials caching tests */
    
  {
	char buf[1500];
	unsigned KRB_INT32 buflen;
	des_cblock sessionKey;
	Key_schedule schedule;

	k_errno = krb_get_ticket_for_service ("rcmd.toad", buf, &buflen,
		0, sessionKey, schedule, "", 0);
	printf ("krb_get_ticket errno = %d (%s)\n", k_errno,
		krb_get_err_text (k_errno));
    }

#endif

    krb_end_session ((char *)NULL);
    return 0;
}

usage()
{
    fprintf(stderr, "Usage: %s [-irvlps] [name][.instance][@REALM]\n", progname);
    exit(1);
}
